### PR TITLE
[WiP] Whisper Implementation

### PR DESCRIPTION
**Pull request by dbogunowicz**

### CODE REVIEW

File: vllm/model_executor/models/whisper.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+        self.key_value_proj_dim = self.d_model
+        self.head_dim = self.d_model // self.total_num_heads
+        if (self.head_dim * self.total_num_heads) != self.d_model:
+            raise ValueError(
+                f"embed_dim must be divisible by num_heads (got `embed_dim`: {self.d_model}"
+                f" and `num_heads`: {num_heads}).")
+
+        self.scaling = self.head_dim**-0.5
+
+        self.k_proj = ColumnParallelLinear(self.d_model,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/whisper.py:63]: Can we use `QKVColumnLinear` for this?

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/whisper.py:63]: Or no because of the bias?

File: vllm/model_executor/models/whisper.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+                               self.d_model,
+                               kernel_size=3,
+                               padding=1)
+        self.conv2 = nn.Conv1d(self.d_model,
+                               self.d_model,
+                               kernel_size=3,
+                               stride=2,
+                               padding=1)
+
+        self.embed_positions = nn.Embedding(self.max_source_positions,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/whisper.py:226]: could this be `VocabParallelEmbedding`?

File: vllm/model_executor/models/whisper.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+
+class WhisperDecoder(nn.Module):
+
+    def __init__(self,
+                 config: WhisperConfig,
+                 linear_method: Optional[LinearMethodBase] = None):
+        super().__init__()
+        self.d_model = config.d_model
+
+        self.embed_tokens = nn.Embedding(config.vocab_size, self.d_model)
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/whisper.py:329]: Can these be `VocabParallelEmbedding`?

File: vllm/model_executor/models/whisper.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+                                          kv_cache=kv_caches,
+                                          input_metadata=input_metadata)
+        return hidden_states
+
+    def sample(self, hidden_states: torch.Tensor,
+               sampling_metadata: SamplingMetadata):
+        # TODO: For now we are not implementing the sampling method
+        return hidden_states
+
+    def load_weights(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/whisper.py:409]: Can you match the style of llama in this function? They have been very consistent with this logic across functions

https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L363

File: vllm/model_executor/models/whisper.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/whisper.py:63]: Maybe we should extend QKVParallelLinear for this case?

Could also do this in a follow up PR

**Review by robertgshaw2-neuralmagic — CHANGES_REQUESTED** [review]
Left a few comments on the model definitions

Will follow up on the core implementation once we synchronize the branches and sync up with the current multimodal implementation upstream

File: vllm/model_executor/models/whisper.py (hunk 2 above)
Comment (dbogunowicz) [reply vllm/model_executor/models/whisper.py:226]: yes, this part is definitely parallelizable, my omission, I did introduce the `VocabParallelEmbedding` here before 🥴

File: vllm/model_executor/models/whisper.py (hunk 3 above)
Comment (dbogunowicz) [reply vllm/model_executor/models/whisper.py:329]: Perhaps, let me check

File: vllm/model_executor/models/whisper.py (hunk 4 above)
Comment (dbogunowicz) [reply vllm/model_executor/models/whisper.py:409]: Looking good, seems like we could adhere to their convention

File: vllm/model_executor/models/whisper.py (hunk 1 above)
Comment (dbogunowicz) [reply vllm/model_executor/models/whisper.py:63]: @robertgshaw2-neuralmagic but nothing prevents us from adding `QKVParallelLinear` support for the T5 model (once again, only for encoder attention and decoder self-attention)!

File: vllm/sequence.py [hunk 5]
```diff
@@ -441,7 +463,9 @@ def __init__(
         self.sampling_params = sampling_params
         self.block_tables = block_tables
         self.lora_request = lora_request
+        self.multi_modal_data = multi_modal_data
         self.prefix = prefix
+        self.multi_modal_data = multi_modal_data
```
Comment (dbogunowicz) [inline vllm/sequence.py:468]:
```suggestion
```

**Comment by andy-neuma** [comment]
stale

**Pull request closed without merging**